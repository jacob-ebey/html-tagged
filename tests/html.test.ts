import { assertSnapshot } from "snapshot";

import { html } from "../mod.ts";

Deno.test("should return an object with an __html property", async (t) => {
  const result = html`<div></div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow string", async (t) => {
  const result = html`<div>${"a"}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow integer", async (t) => {
  const result = html`<div>${1}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow float", async (t) => {
  const result = html`<div>${0.999999999}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow true", async (t) => {
  const result = html`<div>${true}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should not print false", async (t) => {
  const result = html`<div>${false}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should not print null", async (t) => {
  const result = html`<div>${null}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should not print undefined", async (t) => {
  const result = html`<div>${undefined}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow HTMLNode", async (t) => {
  const result = html`<div>${html`<span></span>`}</div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test("should allow array of mixed values", async (t) => {
  const result = html`<div>
    ${["a", 1, true, false, null, undefined, html`<span>2</span>`]}
  </div>`;
  await assertSnapshot(t, result[1]);
});

Deno.test(
  "scripts are found and parsed for bubbling during render",
  async (t) => {
    const result = html`<script>
      console.log("Hello, World!");
    </script>`;
    await assertSnapshot(t, result[2]);
  },
);

Deno.test(
  "nested scripts are found and parsed for bubbling during render",
  async (t) => {
    const result = html`<script>
        console.log("a");
      </script>
      ${html`<script>
        console.log("b");
      </script>`}`;

    await assertSnapshot(t, result[2]);
  },
);

Deno.test(
  "styles are found and parsed for bubbling during render",
  async (t) => {
    const result = html`<style>
      body {
        background-color: red;
      }
    </style>`;
    await assertSnapshot(t, result[2]);
  },
);

Deno.test(
  "nested styles are found and parsed for bubbling during render",
  async (t) => {
    const result = html`<style>
        body {
          background-color: red;
        }
      </style>
      ${html`<style>
        body {
          color: green;
        }
      </style>`}`;
    await assertSnapshot(t, result[2]);
  },
);

Deno.test(
  "custom-elements are found and parsed for expansion during render",
  async (t) => {
    const result = html`<div>
      <custom-element name="value">a</custom-element>
    </div>`;
    await assertSnapshot(t, result[1]);
    await assertSnapshot(t, result[2]);
  },
);

Deno.test(
  "nested custom-elements are found and parsed for expansion during render",
  async (t) => {
    const result = html`<div>
      <custom-element-a name="value">
        a
        <custom-element-b name="value">b</custom-element-b>
      </custom-element-a>
    </div>`;
    await assertSnapshot(t, result[1]);
    await assertSnapshot(t, result[2]);
  },
);

Deno.test("nested custom-elements are found in nested HTMLNode", async (t) => {
  const result = html`<div>
    <custom-element-a name="value">
      a ${html`<custom-element-b name="value">b</custom-element-b>`}
    </custom-element-a>
  </div>`;
  await assertSnapshot(t, result[1]);
  await assertSnapshot(t, result[2]);
});
