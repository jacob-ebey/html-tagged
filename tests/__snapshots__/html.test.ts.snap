export const snapshot = {};

snapshot[`should return an object with an __html property 1`] = `"<div></div>"`;

snapshot[`should allow string 1`] = `"<div>a</div>"`;

snapshot[`should allow integer 1`] = `"<div>1</div>"`;

snapshot[`should allow float 1`] = `"<div>0.999999999</div>"`;

snapshot[`should allow true 1`] = `"<div>true</div>"`;

snapshot[`should not print false 1`] = `"<div></div>"`;

snapshot[`should not print null 1`] = `"<div></div>"`;

snapshot[`should not print undefined 1`] = `"<div></div>"`;

snapshot[`should allow HTMLNode 1`] = `"<div><span></span></div>"`;

snapshot[`should allow array of mixed values 1`] = `
"<div>
    a1true<span>2</span>
  </div>"
`;

snapshot[`scripts are found and parsed for bubbling during render 1`] = `
[
  [
    Symbol(HTMLTag),
    false,
    "script",
    null,
  ],
  '
      console.log("Hello, World!");
    ',
  [
    Symbol(HTMLTag),
    true,
    "script",
    null,
  ],
]
`;

snapshot[`nested scripts are found and parsed for bubbling during render 1`] = `
[
  [
    Symbol(HTMLTag),
    false,
    "script",
    null,
  ],
  '
        console.log("a");
      ',
  [
    Symbol(HTMLTag),
    true,
    "script",
    null,
  ],
  "
      ",
  [
    Symbol(HTMLTag),
    false,
    "script",
    null,
  ],
  '
        console.log("b");
      ',
  [
    Symbol(HTMLTag),
    true,
    "script",
    null,
  ],
]
`;

snapshot[`styles are found and parsed for bubbling during render 1`] = `
[
  [
    Symbol(HTMLTag),
    false,
    "style",
    null,
  ],
  "
      body {
        background-color: red;
      }
    ",
  [
    Symbol(HTMLTag),
    true,
    "style",
    null,
  ],
]
`;

snapshot[`nested styles are found and parsed for bubbling during render 1`] = `
[
  [
    Symbol(HTMLTag),
    false,
    "style",
    null,
  ],
  "
        body {
          background-color: red;
        }
      ",
  [
    Symbol(HTMLTag),
    true,
    "style",
    null,
  ],
  "
      ",
  [
    Symbol(HTMLTag),
    false,
    "style",
    null,
  ],
  "
        body {
          color: green;
        }
      ",
  [
    Symbol(HTMLTag),
    true,
    "style",
    null,
  ],
]
`;

snapshot[`custom-elements are found and parsed for expansion during render 1`] = `
'<div>
      <custom-element name="value">a</custom-element>
    </div>'
`;

snapshot[`custom-elements are found and parsed for expansion during render 2`] = `
[
  "<div>
      ",
  [
    Symbol(HTMLTag),
    false,
    "custom-element",
    'name="value"',
  ],
  "a",
  [
    Symbol(HTMLTag),
    true,
    "custom-element",
    null,
  ],
  "
    </div>",
]
`;

snapshot[`nested custom-elements are found and parsed for expansion during render 1`] = `
"<div>
" +
  '      <custom-element-a name="value">
' +
  "        a
" +
  '        <custom-element-b name="value">b</custom-element-b>
' +
  "      </custom-element-a>
" +
  "    </div>"
`;

snapshot[`nested custom-elements are found and parsed for expansion during render 2`] = `
[
  "<div>
      ",
  [
    Symbol(HTMLTag),
    false,
    "custom-element-a",
    'name="value"',
  ],
  "
        a
        ",
  [
    Symbol(HTMLTag),
    false,
    "custom-element-b",
    'name="value"',
  ],
  "b",
  [
    Symbol(HTMLTag),
    true,
    "custom-element-b",
    null,
  ],
  "
      ",
  [
    Symbol(HTMLTag),
    true,
    "custom-element-a",
    null,
  ],
  "
    </div>",
]
`;

snapshot[`nested custom-elements are found in nested HTMLNode 1`] = `
"<div>
" +
  '    <custom-element-a name="value">
' +
  '      a <custom-element-b name="value">b</custom-element-b>
' +
  "    </custom-element-a>
" +
  "  </div>"
`;

snapshot[`nested custom-elements are found in nested HTMLNode 2`] = `
[
  "<div>
    ",
  [
    Symbol(HTMLTag),
    false,
    "custom-element-a",
    'name="value"',
  ],
  "
      a ",
  [
    Symbol(HTMLTag),
    false,
    "custom-element-b",
    'name="value"',
  ],
  "b",
  [
    Symbol(HTMLTag),
    true,
    "custom-element-b",
    null,
  ],
  "
    ",
  [
    Symbol(HTMLTag),
    true,
    "custom-element-a",
    null,
  ],
  "
  </div>",
]
`;
